python/ffsim/protocols/trace_protocol.py

```diff
@@ -45,3 +51,87 @@ def trace(obj: Any, norb: int, nelec: tuple[int, int]) -> float:
         "The object did not have a _trace_ method that returned the trace "
         "or a _diag_ method that returned its diagonal entries."
     )
+
+
+def _trace_fermion_operator_single(
+    ferm: FermionOperator, norb: int, nelec: tuple[int, int]
+):
+    n_elec_alpha, n_elec_beta = nelec
+    op = list(ferm.keys())[0]
+    coeff = ferm[op]
+    alpha_indices = []
+    beta_indices = []
+    for _, spin, orb in op:
+        if not spin:
+            alpha_indices.append(orb)
+        else:
+            beta_indices.append(orb)
+    alpha_indices = list(set(alpha_indices))
+    beta_indices = list(set(beta_indices))
+    combined_indices = [(i, False) for i in alpha_indices] + [
+        (i, True) for i in beta_indices
+    ]
+
+    no_configuration = False
+    eta_alpha = 0
+    eta_beta = 0
+
+    # loop over the support of the operator
+    # assume that each site is either 0 or 1 at the beginning
+    # track the state of the site through the application of the operator
+    # if the state exceed 1 or goes below 0,
+    # the state is not physical and the trace must be 0
+    for i, spin in combined_indices:
+        initial_zero = 0
+        initial_one = 1
+        is_zero = True
+        is_one = True
+        for action, aspin, orb in reversed(op):
+            if aspin == spin and orb == i:
+                change = action * 2 - 1
+                initial_zero += change
+                initial_one += change
+                if initial_zero < 0 or initial_zero > 1:
+                    is_zero = False
+                if initial_one < 0 or initial_one > 1:
+                    is_one = False
+            if not is_zero and not is_one:
+                no_configuration = True
+                break
+        if (is_zero and initial_zero != 0) or (is_one and initial_one != 1):
+            no_configuration = True
+        # only one case is possible if the operator has support on site i
+        assert not is_zero or not is_one
+        # count the number of electrons
+        if is_one:
+            if not spin:
+                eta_alpha += 1
+            else:
+                eta_beta += 1
+        # the number of electrons exceed the number of allowed electrons
+        if eta_alpha > n_elec_alpha or eta_beta > n_elec_beta:
+            no_configuration = True
+        if no_configuration:
+            trace = 0j
+            break
+
+    if not no_configuration:
+        # the trace is nontrival and is a product of
+        # coeff, and
+        # binom(#orbs not in the support of A, #elec on these orbs)
+        # for each spin species
+
+        trace = coeff * math.comb(norb - len(alpha_indices), n_elec_alpha - eta_alpha)
+        trace = trace * math.comb(norb - len(beta_indices), n_elec_beta - eta_beta)
```

Comment:
This should be written as a single statement. If the line is too long, Ruff should format it automatically, I think.